mlflow/models/evaluation/default_evaluator.py

```diff
@@ -261,28 +274,48 @@ def _do_plot(**kwargs):
 }
 
 
-def _evaluate_custom_metric(index, custom_metric, eval_df, builtin_metrics):
+class _CustomMetric(NamedTuple):
+    """
+    A namedtuple representing a custom metric function and its properties.
+
+    function : the custom metric function
+    name : the name of the custom metric function
+    index : the index of the function in the ``custom_metrics`` argument of mlflow.evaluate
+    artifacts_dir : the path to a temporary directory to store produced artifacts of the function
+    """
+
+    function: Callable
+    name: str
+    index: int
+    artifacts_dir: str
+
+
+def _evaluate_custom_metric(custom_metric_tuple, eval_df, builtin_metrics):
     """
     This function calls the `custom_metric` function and performs validations on the returned
     result to ensure that they are in the expected format. It will raise a MlflowException if
     the result is not in the expected format.
 
-    :param index: The index of the custom metric function in the custom_metrics parameter
-                  of the mlflow.evaluate function.
-    :param custom_metric: A user provided custom metric function.
+    :param custom_metric_tuple: Containing a user provided function and its index in the
+                                ``custom_metrics`` parameter of ``mlflow.evaluate``
     :param eval_df: A Pandas dataframe object containing a prediction and a target column.
     :param builtin_metrics: A dictionary of metrics produced by the default evaluator.
     :return: A tuple of dictionaries. The first is a dictionary of metrics, the second is
              a dictionary of artifacts (which can be None if the custom metric function did
              not produce any).
     """
     exception_header = (
-        f"Custom metric function "
-        f"'{getattr(custom_metric, '__name__', repr(custom_metric))}' at index {index} in the "
-        f"`custom_metrics` parameter"
+        f"Custom metric function '{custom_metric_tuple.name}' at index {custom_metric_tuple.index}"
+        " in the `custom_metrics` parameter"
     )
 
-    result = custom_metric(eval_df, builtin_metrics)
+    if len(inspect.signature(custom_metric_tuple.function).parameters) == 3:
+        result = custom_metric_tuple.function(
+            eval_df, builtin_metrics, custom_metric_tuple.artifacts_dir
```

Comment:
I thought we agreed to not add artifacts_dir to the user-supplied function args. User can treat it as a global variable that always exists during the function runtime. Just note it in the comments sections and examples should be sufficient.